### PR TITLE
new module ts_http_api_SUITE

### PR DESCRIPTION
**Pull request by hmmr**
RTS-258

Originally by @lehoff (submitted in https://github.com/basho/riak_test/pull/1015), with modifications per changes and fixes in https://github.com/basho/riak_kv/pull/1398, now resubmitted against proper branch.

### CODE REVIEW

**Comment by hmmr** [comment]
Now with additional commits with changes to account for a regression in https://github.com/basho/riak_kv/pull/1378, and changes in query return type in https://github.com/basho/riak_kv/pull/1382 post TTB merge. Requires https://github.com/basho/riak_pb/pull/185.

File: tests/ts_simple_describe_table.erl [hunk 1]
```diff
@@ -38,9 +38,9 @@ confirm() ->
...
         {[<<"Column">>,<<"Type">>,<<"Is Null">>,<<"Primary Key">>, <<"Local Key">>],
         [{<<"myfamily">>,  <<"varchar">>,   false,  1,  1},
-        {<<"myseries">>,   <<"varchar">>,   false,  2,  2},
-        {<<"time">>,       <<"timestamp">>, false,  3,  3},
-        {<<"weather">>,    <<"varchar">>,   false, [], []},
-        {<<"temperature">>,<<"double">>,    true,  [], []}]},
+         {<<"myseries">>,   <<"varchar">>,   false,  2,  2},
+         {<<"time">>,       <<"timestamp">>, false,  3,  3},
+         {<<"weather">>,    <<"varchar">>,   false, undefined, undefined},
+         {<<"temperature">>,<<"double">>,    true,  undefined, undefined}]},
```
Comment (lukebakken) [inline tests/ts_simple_describe_table.erl]: I'm assuming the change from `[]` to `undefined` to mean `null` is OK?

Comment (hmmr) [inline tests/ts_simple_describe_table.erl]: I think so. In the original pb terms, `undefined`s are converted to `[]` in riak_pb_ts_codec, iirc.

Comment (lukebakken) [inline tests/ts_simple_describe_table.erl]: Right, but shouldn't this stay the same? I confirmed that `DESCRIBE` does return `undefined` when using TTB encoding at the moment.

Comment (hmmr) [inline tests/ts_simple_describe_table.erl]: Hm. Then it is something that came out unexpectedly from the merge. I'll investigate.

File: tests/ts_http_api_SUITE.erl [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+%%--------------------------------------------------------------------
+%% COMMON TEST CALLBACK FUNCTIONS
+%%--------------------------------------------------------------------
+
+suite() ->
+    [{timetrap,{minutes,10}}].
+
+init_per_suite(Config) ->
+    [_Node|_] = Cluster = ts_util:build_cluster(single),
```
Comment (macintux) [inline tests/ts_http_api_SUITE.erl]: Need to make sure the cluster is fully up and running before `init_per_suite` terminates or else the first test(s) may fail (Brett and I saw this happen).

Suggest `rt:wait_until_nodes_ready/1`.

File: tests/ts_http_api_SUITE.erl [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% Tests for range queries around the boundaries of quanta.
+%%
+%% -------------------------------------------------------------------
+-module(ts_http_api_SUITE).
```
Comment (javajolt) [inline tests/ts_http_api_SUITE.erl]: To match the standard TS test names, this should be `ts_simple_http_api_SUITE`

File: tests/ts_http_api_SUITE.erl [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    {"404", Headers, Body} = list_keys("john", Cfg),
+    "text/plain" = content_type(Headers),
+    "Table \"john\" does not exist" = Body.
+
+%%% select
+select_test(Cfg) ->
+    Select = "select * from bob where a='q1' and b='w1' and c>1 and c<99",
+    {ok,"200", Headers, Body} = execute_query(Select, Cfg),
+    "application/json" = content_type(Headers),
+    io:format("Body ~p\n", [Body]),
```
Comment (javajolt) [inline tests/ts_http_api_SUITE.erl]: Remove debug statement

File: tests/ts_http_api_SUITE.erl [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    "Query error: Unexpected token '''." = Body.
+
+%%% delete
+delete_data_existing_row_test(Cfg) ->
+    {ok, "200", Headers, Body} = delete("bob", "q1", "w1", 11, Cfg),
+    "application/json" = content_type(Headers),
+    Body = success_body(),
+    Select = "select * from bob where a='q1' and b='w1' and c>1 and c<99",
+    {ok, "200", _Headers2,
+     "{\"columns\":[\"a\",\"b\",\"c\",\"d\"],\"rows\":[[\"q1\",\"w1\",20,119]]}"} =
```
Comment (javajolt) [inline tests/ts_http_api_SUITE.erl]: A more flexible JSON parser-based test might be less brittle than an exact string match here

**Comment by javajolt** [comment]
- [ ] Missing single key GET tests

**Comment by hmmr** [comment]
@javajolt There is an implicit test for single-key get, in https://github.com/basho/riak_test/pull/1037/commits/8daf879e7e8cd9af04cb55d68249bd9452a391e1#diff-fab6193d0172e564f9e34f4cda49fefdR189 and the `foreach` loop below.

**Comment by javajolt** [comment]
curl test results:
```
shorty table create (POST) ... ok
shorty insert (POST) ... ok
shorty select (POST) ... ok
shorty list_keys (GET) ... ok
bad query (lexer error) (POST) ... ok
create table (missing key) (POST) ... ok
create table (POST) ... ok
describe table (POST) ... ok
put some data (POST) ... ok
put some bad data (bad key) (POST) ... ok
put some bad data (bad value) (POST) ... ok
put some bad data (missing value for non-null field) (POST) ... ok
single-key get (field-qualified) (GET) ... ok
single-key get (field-qualified, shuffled) (GET) ... ok
single-key non-existing get (GET) ... ok
single-key non-existing get (extra fields) (GET) ... ok
single-key non-existing get (missing fields) (GET) ... ok
single-key get (bad parameter) (GET) ... ok
single-key non-existing delete (DELETE) ... ok
list_keys (GET) ... ok
select query (POST) ... ok
single-key delete (DELETE) ... ok
select query (one record missing) (POST) ... ok
select query (empty) (POST) ... ok
bad get (wrong table) (GET) ... ok
bad put (nxkey) (PUT) ... ok
bad put (wrong table and parameters) (PUT) ... ok
bad query 1 (GET) ... ok
bad query 2 (PUT) ... ok
bad query 3 (POST) ... ok
bad query 4 (POST) ... ok
```

**Comment by javajolt** [comment]
👍 146a632

**Pull request merged**